### PR TITLE
[SPARK-23431][CORE] Expose the new executor memory metrics at the stage level

### PR DESCRIPTION
**Pull request by edwinalu**
## What changes were proposed in this pull request?

Collect and show the new executor memory metrics for each stage, to provide more information on how memory is used per stage. Peak values for metrics are show for each stage. For executor summaries for each stage, the peak values per executor are also shown.

## How was this patch tested?

Added new unit tests.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by felixcheung** [comment]
ok to test

**Comment by squito** [comment]
Jenkins, retest this please

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 1]
```diff
@@ -790,6 +790,19 @@ private[spark] class AppStatusListener(
       }
     }
 
+    // update stage executor metrics
+    event.executorUpdates.foreach { updates =>
+      val activeStages = liveStages.values().asScala
+        .filter(_.status == v1.StageStatus.ACTIVE)
+      activeStages.foreach { stage =>
+        stage.peakExecutorMetrics.compareAndUpdatePeakValues(updates)
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: do you need a `maybeUpdate(stage, now)` after this?

**Review by squito — COMMENTED** [review]
Can you update the PR description to make it clear this is only updating the REST endpoints for now?  As you refer to 'showing' the metrics, it sounds like they're visible in the UI.

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 2]
```diff
@@ -804,12 +817,18 @@ private[spark] class AppStatusListener(
 
   override def onStageExecutorMetrics(executorMetrics: SparkListenerStageExecutorMetrics): Unit = {
     val now = System.nanoTime()
-
+    Option(liveStages.get((executorMetrics.stageId, executorMetrics.stageAttemptId)))
+      .foreach { stage =>
+      stage.peakExecutorMetrics.compareAndUpdatePeakValues(executorMetrics.executorMetrics)
```
Comment (rezasafi) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: nit: Is it possible to change the name of the parameter or the field member? executorMetrics.executorMetrics feels kind of weird

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala [hunk 3]
```diff
@@ -325,6 +325,9 @@ private class LiveExecutorStageSummary(
 
   var metrics = createMetrics(default = 0L)
 
+  // peak values for executor level metrics
+  var peakExecutorMetrics = new ExecutorMetrics()
```
Comment (rezasafi) [inline core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: can this also be a val like the one in line 387?

**Review by rezasafi — CHANGES_REQUESTED** [review]
Thank you very much for the pr.

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala (hunk 3 above)
Comment (edwinalu) [reply core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: Yes, this should be val.

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala (hunk 2 above)
Comment (edwinalu) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: How about "metrics", or "stageExecutorMetrics" for the parameter name?

Comment (rezasafi) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: yeah, sounds good to me.

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 4]
```diff
@@ -802,15 +817,24 @@ private[spark] class AppStatusListener(
...
+        update(stage, now)
+      }
+      val esummary = stage.executorSummary(event.execId)
+      if (esummary.peakExecutorMetrics.compareAndUpdatePeakValues(event.executorMetrics)) {
+        update(esummary, now)
+      }
+    }
     // check if there is a new peak value for any of the executor level memory metrics,
     // while reading from the log. SparkListenerStageExecutorMetrics are only processed
     // when reading logs.
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: I need a little refresher here -- does this comment also apply to the new code you're adding above?

and why again does this only apply for when we're reading from the event logs (maybe this comment should be updated to point to whatever is happening for a live app)

Comment (edwinalu) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: Yes, comments are still applicable, although for stage level. I'll move the comments around, and and add a pointer to onExecutorMetricsUpdate for the live app case.

**Comment by edwinalu** [comment]
@squito and @rezasafi , could you please take another look? Thanks for your review and comments.

**Comment by itamarst** [comment]
Ping—any chance someone could review this patch? These are useful metrics.

**Comment by stackedsax** [comment]
@squito @rezasafi any chance either of you could review this patch?

**Comment by dongjoon-hyun** [comment]
This is reopened according to the community request. I also removed the label, `Stale`.

**Comment by stackedsax** [comment]
@edwinalu thanks for getting this started so long ago.  Looks like there are some conflicts after so much time has passed.  Do you have time/interest to continue working on this or would you like some help getting this sorted out?

**Comment by edwinalu** [comment]
@stackedsax , I don't have time to continue working on this unfortunately. If you or someone else is able to help with adding stage level metrics, that would be great. Please let me know if there's any questions about the original PR.

**Comment by imback82** [comment]
@stackedsax, do you plan to work on this? Otherwise, I can start looking into this next week. Thanks!

**Comment by itamarst** [comment]
@imback82 I'll be taking this up if you don't (on behalf of @stackedsax), but since I am not a Spark contributor at the moment it'd probably be a _lot_ faster and easier if you did it. So please do look into it—thank you!

**Comment by imback82** [comment]
Thanks @itamarst. I started looking into this and will update this thread.

**Comment by imback82** [comment]
I created a PR here: #29020. I will also ask a reviewing committer to put @edwinalu as a co-author if that's the practice when a PR is "inherited". Thanks!

**Comment by stackedsax** [comment]
@imback82 Looks like #29020 is merged into 3.1.0.  Should this issue get closed now, or is it being left open for someone to do the same in 2.x versions of Spark?

**Comment by imback82** [comment]
This can be closed now. We can use changes in #29020 if backporting to 2.4 is needed. (@dongjoon-hyun / @gengliangwang can confirm, but I don't think this will be backported to 2.4.)

**Pull request closed without merging**